libyara/include/yara/types.h

```diff
@@ -827,4 +845,82 @@ struct YR_DICTIONARY_ITEMS
 };
 
 
+
+// Iterators are used in loops of the form:
+//
+// for <any|all|number> <identifier> in <iterator> : ( <expression> )
+//
+// The YR_ITERATOR struct abstracts the many different types of objects that
+// can be iterated. Each type of iterator must provide a "next" function which
+// is called multiple times for retrieving elements from the iterator. This
+// function is responsible for pushing the next item in the stack and a boolean
+// indicating if the end of the iterator has been reached. The boolean must be
+// pushed first, so that the next item is in the top of the stack when the
+// function returns.
+//
+//  +------------+
+//  | next item  |  <- top of the stack
+//  +------------+
+//  | false      |  <- false indicates that there are more items
+//  +------------+
+//  |   . . .    |
+//
+// The boolean shouldn't be true if the next item was pushed in the stack, it
+// can be true only when all the items have been returned in previous calls,
+// in which case the value for the next item should be UNDEFINED. The stack
+// should look like this after the last call to "next":
+//
+//  +------------+
+//  | undefined  |  <- next item is undefined.
+//  +------------+
+//  | true       |  <- true indicates that are no more items.
+//  +------------+
+//  |   . . .    |
+//
+// We can't use the UNDEFINED value in the stack as an indicator of the end
+// of the iterator, because it's legitimate for an iterator to return UNDEFINED
+// items in the middle of the iteration.
+//
+// The "next" function should return ERROR_SUCCESS if every went fine or an
```

Comment:
// The "next" function should return ERROR_SUCCESS if everything went fine or an